axlearn/cloud/gcp/examples/dataflow_inference_custom.py

```diff
@@ -0,0 +1,202 @@
+"""An Apache Beam example pipeline to run batch inference jobs using a model trained with AXLearn.
+
+Command line options:
+--module: the same module used for training
+--config: the same config used for training
+--trainer_dir: location of your checkpoints for inference
+
+To debug locally:
+$ docker run -it --mount type=bind,src=$HOME/.config/gcloud,dst=/root/.config/gcloud \
+    --entrypoint /bin/bash ${DOCKER_REPO}/${DOCKER_IMAGE}:{DOCKER_TAG}
+> python3 -m axlearn.cloud.gcp.examples.dataflow_inference_custom \
+    --module=text.gpt.c4_trainer \
+    --config=fuji-7B-single  \
+    --trainer_dir='gs://.../checkpoints/step_xxx'
+
+To use axlearn CLI:
+$ axlearn gcp dataflow start \
+    --bundler_spec=dockerfile=Dockerfile \
+    --bundler_spec=repo=${DOCKER_REPO} \
+    --bundler_spec=image=${DOCKER_IMAGE} \
+    --bundler_spec=target=dataflow \
+    --bundler_spec=allow_dirty=True \
+    --dataflow_spec=runner=DataflowRunner \
+    -- "'python3 -m axlearn.cloud.gcp.examples.dataflow_inference_custom \
+        --module=text.gpt.c4_trainer \
+        --config=fuji-7B-single \
+        --trainer_dir='gs://.../checkpoints/step_xxx' \
+        '"
+
+To use GPUs for your job:
+$ axlearn gcp dataflow start \
+    --bundler_spec=dockerfile=Dockerfile \
+    --bundler_spec=repo=${DOCKER_REPO} \
+    --bundler_spec=image=${DOCKER_IMAGE} \
+    --bundler_spec=target=dataflow \
+    --bundler_spec=allow_dirty=True \
+    --dataflow_spec=runner=DataflowRunner \
+    --dataflow_spec=dataflow_service_options=\
+    "worker_accelerator=type:nvidia-l4;count:1;install-nvidia-driver" \
+    -- "'python3 -m axlearn.cloud.gcp.examples.dataflow_inference_custom \
+        --module=text.gpt.c4_trainer \
+        --config=fuji-7B-single \
+        --trainer_dir='gs://.../checkpoints/step_xxx' \
+        '"
+
+"""
+
+
+import copy
+import logging
+from typing import Any, Dict, Optional, Sequence
+
+import apache_beam as beam
+import jax
+from absl import app, flags
+from absl.flags import argparse_flags
+from apache_beam.ml.inference.base import ModelHandler, PredictionResult, RunInference
+from apache_beam.options.pipeline_options import PipelineOptions
+
+import axlearn.common.input_fake as input_fake
+import axlearn.common.launch_trainer as trainer_utils
+from axlearn.common.inference import InferenceRunner, MethodRunner
+from axlearn.common.utils import NestedTensor
+
+
+class CustomModelHandler(ModelHandler[Dict, PredictionResult, Any]):
```

Comment:
Looks like this is a new thing. Do you have document pointer for this ModelHandler ?

Reply:
@yqwangustc Yes RunInference is fairly new but can be useful in handling throughput, etc. It natively supports TF, PyTorch, HF pre-trained models, but for JAX (axlearn-trained) models, we can use a custom model handler: https://cloud.google.com/dataflow/docs/notebooks/run_custom_inference

Reply:
Done 👍🏼 I think all comments are resolved now @ruomingp @markblee